### PR TITLE
Made the logic in the files contact_book and record

### PR DESCRIPTION
**Pull request by Chonna86**
1. In the file сontact_bot, he created the AddressBook class, which is copied from UserDict, and wrote all the necessary methods of this class
2. In the record file, I wrote all the necessary classes with methods
3. Handlers added

### CODE REVIEW

File: src/user_actions_handler.py
Comment (filiurskyi) [inline src/user_actions_handler.py:1]: i think these imports 
```
from personal_assistant.src.address_book import contact_book
from personal_assistant.src.address_book.contact_book import AddressBook, Record
```

should be like 
```
from address_book import contact_book
from address_book.contact_book import AddressBook, Record
```

File: src/address_book/contact_book.py [hunk 1]
```diff
@@ -1,5 +1,47 @@
 from collections import UserDict
+from datetime import datetime,timedelta
+import pickle
+from src.record import Record
```
Comment (filiurskyi) [inline src/address_book/contact_book.py]: here simple import as well as follows:

`from record import Record`

File: src/user_actions_handler.py [hunk 2]
```diff
@@ -29,9 +41,92 @@ def handler_bye(*args):
...
+        return "No matching contacts found."
+
 
 OPERATORS = {
     'hello': handler_greetings,
     'close': handler_bye,
     'exit': handler_bye,
-}
+    'show all': show_all_contacts,
+    'add': add_contact,
```
Comment (filiurskyi) [inline src/user_actions_handler.py]: `'add contact': add_contact,`

**Review by filiurskyi — CHANGES_REQUESTED** [review]
> Please notice that contents of commands...
> 
> in parser.py:
> `COMMANDS_RGX`
> 
> and user_actions_handler.py:
> `OPERATORS`
> 
> should bo equal
is already in work via  globals.py

After issues below are resolved, the whole code should work.
Next step will be to run main.py and test each function of bot.

File: src/user_actions_handler.py [hunk 3]
```diff
@@ -29,9 +41,92 @@ def handler_bye(*args):
...
+def show_all_contacts(*args):
+    if not contact_book.data:
+        return "No contacts found"
+    contact_list = "\n".join([str(record) for record in contact_book.data.values()])
+    return contact_list
+
+# Function to add a new contact
+@input_error
+def add_contact(*args):
+    command = " ".join(args)
```
Comment (filiurskyi) [inline src/user_actions_handler.py]: no empty input check

File: src/record.py [hunk 4]
```diff
@@ -3,16 +3,86 @@
...
 
 class Phone(Field):
-    pass
+    def __init__(self, value):
+        super().__init__(value)
+        self.validate_phone(value)
+
+    def validate_phone(self, phone):
+        if not (len(phone) == 10 and phone.isdigit()):
+            raise ValueError("Invalid phone number format")
```
Comment (filiurskyi) [inline src/record.py]: Error rised but nothing in @input_handler decorator

File: src/user_actions_handler.py [hunk 5]
```diff
@@ -29,9 +41,92 @@ def handler_bye(*args):
...
+@input_error
+def show_all_contacts(*args):
+    if not contact_book.data:
+        return "No contacts found"
+    contact_list = "\n".join([str(record) for record in contact_book.data.values()])
+    return contact_list
+
+# Function to add a new contact
+@input_error
+def add_contact(*args):
```
Comment (filiurskyi) [inline src/user_actions_handler.py]: We call OPERATORS with 1 arg, no need to use *args. Parser calls function with only one parameter of type list.
So i think it is better to do this:
`def add_contact(user_input):`
and rename `args `into `user_input`

File: src/user_actions_handler.py [hunk 6]
```diff
@@ -29,9 +41,92 @@ def handler_bye(*args):
 def get_handler(operator):
     return OPERATORS[operator]
 
+def save_address_book():
+    with open(file, "wb") as fh:
+        pickle.dump(contact_book, fh)
+
+# Function to display all contacts
+@input_error
+def show_all_contacts(*args):
```
Comment (filiurskyi) [inline src/user_actions_handler.py]: same problem as `def add_contact`

File: src/user_actions_handler.py [hunk 7]
```diff
@@ -29,9 +41,92 @@ def handler_bye(*args):
...
+        return "Give me name, phone, and birthday please"
+    name, phone, birthday = args
+    record = Record(name, phone, birthday)
+    contact_book.add_record(record)
+    save_address_book()
+    return f"Contact {name} added with phone number {phone} and birthday {birthday}"
+
+# Function to change a contact's phone number
+@input_error
+def change_phone(*args):
```
Comment (filiurskyi) [inline src/user_actions_handler.py]: same proble as `def add_contact`

File: src/user_actions_handler.py [hunk 8]
```diff
@@ -29,9 +41,92 @@ def handler_bye(*args):
...
+    if record:
+        record.edit_phone(record.phones[0].value, new_phone)
+        save_address_book()
+        return f"Phone number for {name} updated to {new_phone}"
+    else:
+        return "Contact not found"
+
+# Function to get a contact's phone number
+@input_error
+def get_phone(*args):
```
Comment (filiurskyi) [inline src/user_actions_handler.py]: same proble as `def add_contact`

File: src/user_actions_handler.py [hunk 9]
```diff
@@ -29,9 +41,92 @@ def handler_bye(*args):
...
+    _, name = command.split()
+    record = contact_book.find(name)
+    if record:
+        return f"Phone number for {name}: {record.phones[0].value}"
+    else:
+        return "Contact not found"
+
+# Function to get a contact's birthday
+@input_error
+def get_birthday(*args):
```
Comment (filiurskyi) [inline src/user_actions_handler.py]: same proble as `def add_contact`

File: src/user_actions_handler.py (hunk 3 above)
Comment (filiurskyi) [inline src/user_actions_handler.py]: function recieves list (examlple: `['John', '1234567898', '1999-05-05']`)
no need to work with args, logic has to work with this given input

Comment (filiurskyi) [reply src/user_actions_handler.py]: you can use this instead to split input into needed arguments:

```
    name = args[0]
    phone = args[1]
    birthday = args[2]
```

File: src/user_actions_handler.py [hunk 10]
```diff
@@ -29,9 +41,92 @@ def handler_bye(*args):
 def get_handler(operator):
     return OPERATORS[operator]
 
+def save_address_book():
```
Comment (filiurskyi) [inline src/user_actions_handler.py]: we do not need to save here as we already save to pickle in `main.py `on each user command

File: src/user_actions_handler.py [hunk 11]
```diff
@@ -29,9 +41,92 @@ def handler_bye(*args):
...
+# Function to add a new contact
+@input_error
+def add_contact(*args):
+    command = " ".join(args)
+    _, *args = command.split(maxsplit=3)
+    if len(args) < 3:
+        return "Give me name, phone, and birthday please"
+    name, phone, birthday = args
+    record = Record(name, phone, birthday)
+    contact_book.add_record(record)
```
Comment (filiurskyi) [inline src/user_actions_handler.py]: contact_book is wrong. 

we already made instance of Addressbook in line 13:
`book = AddressBook()`

so use:
`book.add_record(record)`

File: src/address_book/contact_book.py [hunk 12]
```diff
@@ -1,5 +1,47 @@
...
+    def delete(self, name):
+        if name in self.data:
+            del self.data[name]
+
+    def iterator(self, N=1):
+        records = list(self.data.values())
+        for i in range(0, len(records), N):
+            yield records[i:i + N]
+
+    def save_to_file(self, file_path):
```
Comment (Baskina) [inline src/address_book/contact_book.py]: Delete this method please. Files are saved automatically.

File: src/address_book/contact_book.py [hunk 13]
```diff
@@ -1,5 +1,47 @@
...
+        with open(file_path, 'wb') as file:
+            pickle.dump(self.data, file)
+
+    
+    def load_from_file(self, file_path):
```
Comment (Baskina) [inline src/address_book/contact_book.py]: the same. No needs in this method

File: src/record.py [hunk 14]
```diff
@@ -3,16 +3,86 @@
...
+    @property
+    def value(self):
+        return self._value
 
+    @value.setter
+    def value(self, new_value):
+        self._value = new_value
 
 class Name(Field):
     pass
```
Comment (Baskina) [inline src/record.py]: Please add logic to this class

File: src/user_actions_handler.py (hunk 6 above)
Comment (Baskina) [inline src/user_actions_handler.py]:
```suggestion
def handler_show_all_contacts(*args):
```

File: src/user_actions_handler.py (hunk 5 above)
Comment (Baskina) [inline src/user_actions_handler.py]:
```suggestion
def handler_add_contact(*args):
```

File: src/user_actions_handler.py [hunk 15]
```diff
@@ -29,9 +41,92 @@ def handler_bye(*args):
...
+    save_address_book()
```
Comment (Baskina) [inline src/user_actions_handler.py]: No needs in this method call

File: src/user_actions_handler.py (hunk 10 above)
Comment (Baskina) [inline src/user_actions_handler.py]: Delete this method please

File: src/user_actions_handler.py [hunk 16]
```diff
@@ -29,9 +41,92 @@ def handler_bye(*args):
...
+    else:
+        return "No matching contacts found."
+
 
 OPERATORS = {
     'hello': handler_greetings,
     'close': handler_bye,
     'exit': handler_bye,
-}
+    'show all': show_all_contacts,
```
Comment (Baskina) [inline src/user_actions_handler.py]: please change each handler's according to this template: '**handler**_name_of_function'

File: src/user_actions_handler.py (hunk 11 above)
Comment (Baskina) [inline src/user_actions_handler.py]: all methods should be taken from `book`

File: src/user_actions_handler.py [hunk 17]
```diff
@@ -11,12 +12,23 @@
 except FileNotFoundError:
     book = AddressBook()
 
+ 
+def input_error(func):
```
Comment (Baskina) [inline src/user_actions_handler.py]: no need to create separate input_error. Please use one that already in the code (file: error_handler.py), but extend it according to your errors.

File: src/user_actions_handler.py [hunk 18]
```diff
@@ -77,5 +166,10 @@ def get_handler(operator):
     "exit": handler_bye,
     "good bye": handler_bye,
     "sort dir": handler_sort,
-    "add contact": handler_add_contact
+    "show all contacts": show_all_contacts,
```
Comment (Baskina) [inline src/user_actions_handler.py]: update all naming for hadlers based in template:
```
handler_bye
handler_sort
handler_add_contact
```

File: src/address_book/contact_book.py (hunk 12 above)
Comment (Chonna86) [reply src/address_book/contact_book.py]: Ок

File: src/address_book/contact_book.py (hunk 13 above)
Comment (Chonna86) [reply src/address_book/contact_book.py]: Ок

File: src/record.py (hunk 14 above)
Comment (Chonna86) [reply src/record.py]: Name class in my program can remain unchanged. I've already inherited the Field class for it and don't need any additional methods or properties for the Name class. The Name class is used to represent the name of a contact and stores its value. My code already includes all the provisioning methods from the Field class, such as __str__, @property, @value.setter, as the Field class defines the general logic to manage the value of the field.

File: src/user_actions_handler.py (hunk 17 above)
Comment (Chonna86) [reply src/user_actions_handler.py]: was fixed

File: src/user_actions_handler.py (hunk 10 above)
Comment (Chonna86) [reply src/user_actions_handler.py]: Ok

File: src/user_actions_handler.py (hunk 6 above)
Comment (Chonna86) [reply src/user_actions_handler.py]: Ok

File: src/user_actions_handler.py (hunk 5 above)
Comment (Chonna86) [reply src/user_actions_handler.py]: Ok

File: src/user_actions_handler.py (hunk 15 above)
Comment (Chonna86) [reply src/user_actions_handler.py]: Ok

File: src/user_actions_handler.py (hunk 18 above)
Comment (Chonna86) [reply src/user_actions_handler.py]: Ok

File: src/user_actions_handler.py (hunk 11 above)
Comment (Chonna86) [reply src/user_actions_handler.py]: changed

**Comment by filiurskyi** [comment]
updated code, please check.

**Review by filiurskyi — APPROVED** [review]

**Review by Baskina — APPROVED** [review]

**Pull request merged**